### PR TITLE
MySQL support for Nessie(JDBC based)

### PR DESCRIPTION
**Pull request by vyj7**
Added support to use mySQL database for nessie.

### CODE REVIEW

**Comment by adutra** [comment]
Hi @vyj7 thanks for contributing to Nessie! It turns out, I was myself planning to introduce support for MySQL. I'm afraid this will require a lot more changes compared to what you have in this PR. Are you willing to continue working on this (in which case, I could give you some pointers), or should I take it over from here? Just le me know :-) Thanks!

(Note: you also would need to sign our CLA.)

**Comment by vyj7** [comment]
@adutra Yes I would love to work on this. Can you guide me what else I am missing.If you can give me a small walkthrough or doc how code is structured i can pick up

**Comment by adutra** [comment]
> @adutra Yes I would love to work on this. Can you guide me what else I am missing.If you can give me a small walkthrough or doc how code is structured i can pick up

@vyj7 awesome, but we need to sort out a few legal issues first. I will ping you when we're ready.

**Comment by adutra** [comment]
Hi @vyj7 we're ready to work on this. 

We decided that it's not safe for Nessie to be distributed with the MySQL driver. But there is one thing we can do: offer support for MariaDB instead, and use the MariaDB driver to connect to either MariaDB or MySQL.

So, if you are still OK to work on this, here are the first steps:

1. In `gradle/libs.versions.toml`, declare the MariaDB driver:
```toml
mariadb-java-client = { module = "org.mariadb.jdbc:mariadb-java-client", version = "3.3.3" }
```

2. In `gradle/license/allowed-licenses.json`, declare the LGPL license:
```json
    {
      "moduleLicense": "GNU Lesser General Public License Version 2.1"
    },
```

3. In `gradle/license/allowed-licenses.json` add the following:
```json
    { "bundleName" : "lgpl21", "licenseNamePattern" : "LGPL-2.1" },
```

4. In `versioned/storage/jdbc/build.gradle.kts`, declare the MariaDB driver:
```kotlin
  intTestRuntimeOnly(libs.mariadb.java.client)
```

5. Add support for MariaDB/MySQL in `DatabaseSpecifics.MariaDBDatabaseSpecific` – since I already worked on this a bit, I'm giving you my version. Note: I used the `utf8mb4_bin` collation because it works for both MySQL and MariaDB, but if you have a better one, no problem. Here it is:
```java
  static class MariaDBDatabaseSpecific implements DatabaseSpecific {

    private static final String VARCHAR = "VARCHAR(255) CHARACTER SET utf8mb4 COLLATE utf8mb4_bin";
    private static final String TEXT = "TEXT CHARACTER SET utf8mb4 COLLATE utf8mb4_bin";
    private static final String MYSQL_CONSTRAINT_VIOLATION_SQL_STATE = "23000";

    private final Map<JdbcColumnType, String> typeMap;
    private final Map<JdbcColumnType, Integer> typeIdMap;

    MariaDBDatabaseSpecific() {
      typeMap = new EnumMap<>(JdbcColumnType.class);
      typeIdMap = new EnumMap<>(JdbcColumnType.class);
      typeMap.put(JdbcColumnType.NAME, VARCHAR);
      typeIdMap.put(JdbcColumnType.NAME, Types.VARCHAR);
      typeMap.put(JdbcColumnType.OBJ_ID, VARCHAR);
      typeIdMap.put(JdbcColumnType.OBJ_ID, Types.VARCHAR);
      typeMap.put(JdbcColumnType.OBJ_ID_LIST, TEXT);
      typeIdMap.put(JdbcColumnType.OBJ_ID_LIST, Types.VARCHAR);
      typeMap.put(JdbcColumnType.BOOL, "BIT(1)");
      typeIdMap.put(JdbcColumnType.BOOL, Types.BIT);
      typeMap.put(JdbcColumnType.VARBINARY, "BLOB");
      typeIdMap.put(JdbcColumnType.VARBINARY, Types.BLOB);
      typeMap.put(JdbcColumnType.BIGINT, "BIGINT");
      typeIdMap.put(JdbcColumnType.BIGINT, Types.BIGINT);
      typeMap.put(JdbcColumnType.VARCHAR, VARCHAR);
      typeIdMap.put(JdbcColumnType.VARCHAR, Types.VARCHAR);
    }

    @Override
    public Map<JdbcColumnType, String> columnTypes() {
      return typeMap;
    }

    @Override
    public Map<JdbcColumnType, Integer> columnTypeIds() {
      return typeIdMap;
    }

    @Override
    public boolean isConstraintViolation(SQLException e) {
      return MYSQL_CONSTRAINT_VIOLATION_SQL_STATE.equals(e.getSQLState());
    }

    @Override
    public String wrapInsert(String sql) {
      return sql.replace("INSERT INTO", "INSERT IGNORE INTO");
    }
  }
```

Once you're done we can tackle the tests. Good luck :-)

**Comment by vyj7** [comment]
@adutra  yes I think mariaDB driver should also work for mysql. I will work on this.

**Comment by adutra** [comment]
@vyj7 I think you accidentally closed this PR?

**Comment by vyj7** [comment]
I removed changes done for mysql. I will push changes on this branch only.

**Comment by vyj7** [comment]
@adutra I couldn't find gradle/license/allowed-licenses.json and gradle/license/allowed-licenses.json in project

**Comment by adutra** [comment]
> @adutra I couldn't find gradle/license/allowed-licenses.json and gradle/license/allowed-licenses.json in project

@vyj7 it's because your branch is out of date with the `main` branch. Let's ignore that for now and focus on the rest.

**Comment by adutra** [comment]
@vyj7 for the tests, here's what you need to do:

1. In `nessie-versioned-storage-jdbc-tests` implement `MariaDBBackendTestFactory` et `MySQLBackendTestFactory`.
2. In `nessie-versioned-storage-jdbc` implement the following tests:
  a.  `ITMariaDBBackendFactory`
  b. `ITMariaDBPersist`
  c. `ITMariaDBVersionStore`
  d. `ITMySQLBackendFactory`
  e. `ITMySQLPersist`
  f. `ITMySQLVersionStore`

File: gradle/libs.versions.toml [hunk 1]
```diff
@@ -145,6 +145,7 @@ undertow-servlet = { module = "io.undertow:undertow-servlet", version.ref = "und
 vertx-core = { module = "io.vertx:vertx-core", version = "4.5.7" }
 weld-se-core = { module = "org.jboss.weld.se:weld-se-core", version = "5.1.2.Final" }
 wiremock = { module = "com.github.tomakehurst:wiremock-jre8-standalone", version = "2.35.2" }
+mariadb-java-client = { module = "org.mariadb.jdbc:mariadb-java-client", version = "3.3.3" }
```
Comment (adutra) [inline gradle/libs.versions.toml:148]: Please keep the alphabetical order in this file.

File: versioned/storage/jdbc/src/main/java/org/projectnessie/versioned/storage/jdbc/DatabaseSpecifics.java [hunk 2]
```diff
@@ -141,4 +141,63 @@ public String wrapInsert(String sql) {
       return sql + " ON CONFLICT DO NOTHING";
     }
   }
+
+  static class MariaDBDatabaseSpecific implements DatabaseSpecific {
```
Comment (adutra) [inline versioned/storage/jdbc/src/main/java/org/projectnessie/versioned/storage/jdbc/DatabaseSpecifics.java:145]: You also have to modify the `detect` method and add 2 more case branches: `"mariadb"` and `"mysql"`.

**Comment by snazy** [comment]
> @vyj7 for the tests, here's what you need to do:

Also:
`ITRestApiPersistMariaDB` in `nessie-quarkus`
`QuarkusTestProfilePersistMariaDB` + `MariaDBTestResourceLifecycleManager` in `nessie-quarkus-tests`

File: versioned/storage/jdbc/src/main/java/org/projectnessie/versioned/storage/jdbc/DatabaseSpecifics.java (hunk 2 above)
Comment (vyj7) [reply versioned/storage/jdbc/src/main/java/org/projectnessie/versioned/storage/jdbc/DatabaseSpecifics.java:145]: yes i have added in subsequent change

**Comment by adutra** [comment]
> > @vyj7 for the tests, here's what you need to do:
> 
> Also: `ITRestApiPersistMariaDB` in `nessie-quarkus` `QuarkusTestProfilePersistMariaDB` + `MariaDBTestResourceLifecycleManager` in `nessie-quarkus-tests`

Not in this PR please :-) Let's tackle just the persist layer, the rest is much more involved and will be done in a follow-up PR.

File: gradle/libs.versions.toml [hunk 3]
```diff
@@ -104,7 +104,12 @@ junit-platform-reporting = { module = "org.junit.platform:junit-platform-reporti
 logback-classic = { module = "ch.qos.logback:logback-classic", version.ref = "logback" }
 keycloak-admin-client = { module = "org.keycloak:keycloak-admin-client", version.ref = "keycloak" }
 maven-resolver-supplier = { module = "org.apache.maven.resolver:maven-resolver-supplier", version.ref = "mavenResolver" }
+<<<<<<< feat/mysql_support
```
Comment (adutra) [inline gradle/libs.versions.toml:107]: Heads up: you forgot to solve conflicts in this file.

File: servers/quarkus-common/src/main/java/org/projectnessie/quarkus/providers/storage/JdbcBackendBuilder.java [hunk 4]
```diff
@@ -44,11 +44,14 @@ public class JdbcBackendBuilder implements BackendBuilder {
 
   @Override
   public Backend buildBackend() {
-    if (!DatabaseKind.isPostgreSQL(databaseKind) && !DatabaseKind.isH2(databaseKind)) {
+    if (!DatabaseKind.isPostgreSQL(databaseKind)
```
Comment (adutra) [inline servers/quarkus-common/src/main/java/org/projectnessie/quarkus/providers/storage/JdbcBackendBuilder.java:47]: Please revert this change; let's not touch the quarkus modules, otherwise this PR will become a huge change.

File: servers/quarkus-common/build.gradle.kts [hunk 5]
```diff
@@ -45,6 +45,7 @@ dependencies {
   implementation("io.quarkus:quarkus-hibernate-validator")
   implementation("io.quarkus:quarkus-agroal")
   implementation("io.quarkus:quarkus-jdbc-postgresql")
+  implementation("io.quarkus:quarkus-jdbc-mariadb")
```
Comment (adutra) [inline servers/quarkus-common/build.gradle.kts:48]: Please also revert this one.

File: servers/quarkus-common/src/main/java/org/projectnessie/quarkus/providers/storage/JdbcBackendBuilder.java (hunk 4 above)
Comment (vyj7) [reply servers/quarkus-common/src/main/java/org/projectnessie/quarkus/providers/storage/JdbcBackendBuilder.java:47]: ok i thought this might be needed as well

**Comment by vyj7** [comment]
@adutra  how should i run this code in order to test it with a local mysql/maria-Db image.
I changed properties `servers/quarkus-server/src/main/resources/application.properties`
`nessie.version.store.type=JDBC
quarkus.datasource.db-kind=mariadb
quarkus.datasource.username=user
quarkus.datasource.password=password
quarkus.datasource.jdbc.url=jdbc:mariadb://localhost:3302/my_database`

and then ran task in nessie Quarkus - quarkusRun

**Comment by adutra** [comment]
@vyj7 you don't need to run the server, all you need to do is run the JDBC persistence tests:

```
./gradlew :nessie-versioned-storage-jdbc:build
```

If the build passes once you've added the new tests, then you are done :-)

File: gradle/libs.versions.toml [hunk 6]
```diff
@@ -104,7 +104,8 @@ junit-platform-reporting = { module = "org.junit.platform:junit-platform-reporti
 logback-classic = { module = "ch.qos.logback:logback-classic", version.ref = "logback" }
 keycloak-admin-client = { module = "org.keycloak:keycloak-admin-client", version.ref = "keycloak" }
 maven-resolver-supplier = { module = "org.apache.maven.resolver:maven-resolver-supplier", version.ref = "mavenResolver" }
-micrometer-core = { module = "io.micrometer:micrometer-core", version = "1.13.0" }
+mariadb-java-client = { module = "org.mariadb.jdbc:mariadb-java-client", version = "3.3.3" }
+micrometer-core = { module = "io.micrometer:micrometer-core", version = "1.12.5" }
```
Comment (adutra) [inline gradle/libs.versions.toml:108]: Beware: you inadvertently downgraded this dependency. Please revert to 1.13.0 as before.

**Comment by vyj7** [comment]
@adutra  I have committed the TCs. All the testcases actually failed for mariaDB(TCs are failing) and mySQL(driver not found). I am debugging what are the causes of it. So far the test containers are spawning correctly.

**Comment by vyj7** [comment]
@adutra  Caused by: java.lang.ClassNotFoundException: org.projectnessie.nessie.relocated.protobuf.UnsafeByteOperations
Do you know about this error? I directly ran one of the integration test to see if changes work

**Comment by adutra** [comment]
> @adutra Caused by: java.lang.ClassNotFoundException: org.projectnessie.nessie.relocated.protobuf.UnsafeByteOperations Do you know about this error? I directly ran one of the integration test to see if changes work

Yes this happens sometimes in your IDE. You need to build the `nessie-protobuf-relocated` module:

```
 ./gradlew :nessie-protobuf-relocated:build
```

File: versioned/storage/jdbc-tests/src/main/java/org/projectnessie/versioned/storage/jdbctests/MySQLBackendTestFactory.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  }
+
+  @Nonnull
+  @Override
+  protected JdbcDatabaseContainer<?> createContainer() {
+    String dockerImage = dockerImage("mysql");
+    return new MariaDBDriverMySQLContainer(dockerImage);
+  }
+
+  /** This has been done so that mariaDB Driver can be used for mysql Testcontainer */
```
Comment (adutra) [inline versioned/storage/jdbc-tests/src/main/java/org/projectnessie/versioned/storage/jdbctests/MySQLBackendTestFactory.java:37]: You have a checkstyle violation here:

```
MySQLBackendTestFactory.java:37: First sentence of Javadoc is missing an ending period. [SummaryJavadoc]
```

In fact I think you can remove this javadoc completely as the class code itself is quite self-explanatory.

**Review by adutra — COMMENTED** [review]
@vyj7 You have a few test failures in `AbstractTestJdbcBackendFactory.incompatibleTableSchema()`.

I think you can easily fix them by replacing all occurrences of `VARCHAR` with `VARCHAR(255)`.

**Comment by vyj7** [comment]
@adutra All testcases passed after the changes

**Review by adutra — APPROVED** [review]
Thank you so much for your contribution @vyj7 !
